src/coreclr/src/gc/gc.cpp

```diff
@@ -28045,182 +27942,173 @@ void gc_heap::revisit_written_pages (BOOL concurrent_p, BOOL reset_only_p)
-                    high_address = min (high_address, background_saved_highest_address);
-                }
-                else
-                {
-                    high_address = high_page (seg, concurrent_p);
+                    base_address = max (base_address, background_saved_lowest_address);
+                    dprintf (3, ("h%d: reset only starting %Ix", heap_number, base_address));
                 }
 
-                if ((base_address < high_address) &&
-                    (bcount >= array_size))
-                {
-                    ptrdiff_t region_size = high_address - base_address;
-                    dprintf (3, ("h%d: gw: [%Ix(%Id)", heap_number, (size_t)base_address, (size_t)region_size));
+                dprintf (3, ("h%d: starting: %Ix, seg %Ix-%Ix", heap_number, base_address, 
+                    heap_segment_mem (seg), heap_segment_reserved (seg)));
 
-#ifdef FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP
-                    // When the runtime is not suspended, it's possible for the table to be resized concurrently with the scan
-                    // for dirty pages below. Prevent that by synchronizing with grow_brick_card_tables(). When the runtime is
-                    // suspended, it's ok to scan for dirty pages concurrently from multiple background GC threads for disjoint
-                    // memory regions.
-                    if (!is_runtime_suspended)
+
+                while (1)
+                {
+                    if (reset_only_p)
                     {
-                        enter_spin_lock(&gc_lock);
+                        high_address = ((seg == ephemeral_heap_segment) ? alloc_allocated : heap_segment_allocated (seg));
+                        high_address = min (high_address, background_saved_highest_address);
                     }
-#endif // FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP
-
-                    get_write_watch_for_gc_heap (reset_watch_state, base_address, region_size,
-                                                 (void**)background_written_addresses,
-                                                 &bcount, is_runtime_suspended);
-
-#ifdef FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP
-                    if (!is_runtime_suspended)
+                    else
                     {
-                        leave_spin_lock(&gc_lock);
+                        high_address = high_page (seg, concurrent_p);
                     }
-#endif // FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP
 
-                    if (bcount != 0)
+                    if ((base_address < high_address) &&
+                        (bcount >= array_size))
                     {
-                        total_dirtied_pages += bcount;
+                        ptrdiff_t region_size = high_address - base_address;
+                        dprintf (3, ("h%d: gw: [%Ix(%Id)", heap_number, (size_t)base_address, (size_t)region_size));
+
+    #ifdef FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP
```

Comment:
nit: I know you just copied this code with a different indentation but #ifdef's always start at the beginning of a line.